nsls2forge_utils/git_utils.py

```diff
@@ -0,0 +1,163 @@
+'''
+This is a rework of
+https://github.com/regro/cf-scripts/blob/31ffe0333e9b53d4bce03818c5dee3f08559baea/conda_forge_tick/git_utils.py
+Some of the code was not importable so the
+function had to be copied and reimplemented here
+'''
+import time
+import os
+
+import github3
+from conda_forge_tick.xonsh_utils import indir, env
+from conda_forge_tick.git_utils import feedstock_repo
+from doctr.travis import run_command_hiding_token as doctr_run
+
+
+def fork_url(feedstock_url, username, organization='nsls-ii-forge'):
+    """Creates the URL of the user's fork."""
+    beg, end = feedstock_url.rsplit("/", 1)
+    beg = beg[:-len(organization)]  # chop off organization
+    url = beg + username + "/" + end
+    return url
+
+
+def feedstock_url(fctx, organization='nsls-ii-forge', protocol="ssh"):
+    """Returns the URL for a conda-forge feedstock."""
+    feedstock = fctx.feedstock_name + "-feedstock"
+    if feedstock.startswith("http://github.com/"):
+        return feedstock
+    elif feedstock.startswith("https://github.com/"):
+        return feedstock
+    elif feedstock.startswith("git@github.com:"):
+        return feedstock
+    protocol = protocol.lower()
+    if protocol == "http":
+        url = f"http://github.com/{organization}/" + feedstock + ".git"
+    elif protocol == "https":
+        url = f"https://github.com/{organization}/" + feedstock + ".git"
+    elif protocol == "ssh":
+        url = f"git@github.com:{organization}/" + feedstock + ".git"
+    else:
+        msg = "Unrecognized github protocol {0!r}, must be ssh, http, or https."
+        raise ValueError(msg.format(protocol))
+    return url
+
+
+def get_repo(ctx, fctx, branch, organization='nsls-ii-forge', feedstock=None,
+             protocol="ssh", pull_request=True, fork=True):
+    """
+    Get the feedstock repo from the specified GitHub organization
+
+    Parameters
+    ----------
+    feedstock: str, optional
+        The feedstock to clone if None use $FEEDSTOCK
+    protocol: str, optional
+        The git protocol to use, defaults to ``ssh``
+    pull_request: bool, optional
+        If true issue pull request, defaults to true
+    fork: bool
+        If true create a fork, defaults to true
+    gh: github3.GitHub instance, optional
+        Object for communicating with GitHub, if None build from $USERNAME
+        and $PASSWORD, defaults to None
+
+    Returns
+    -------
+    recipe_dir : str
+        The recipe directory
+    """
+    gh = ctx.gh
+    # first, let's grab the feedstock locally
+    upstream = feedstock_url(fctx=fctx, protocol=protocol)
+    origin = fork_url(upstream, ctx.github_username)
+    feedstock_reponame = feedstock_repo(fctx=fctx)
+
+    if pull_request or fork:
+        repo = gh.repository(organization, feedstock_reponame)
+        if repo is None:
+            fctx.attrs["bad"] = f"{fctx.package_name}: does not match feedstock name\n"
+            return False
```

Comment:
This does not match the return type in the docstring. Should it raise an exception instead?

Reply:
I believe we don't want to stop execution if one of the packages are bad. I think returning `None` here instead is a better option. I have to make sure this doesn't break any of the `conda_forge_tick` checks though.